lms/djangoapps/program_enrollments/link_program_enrollments.py

```diff
@@ -0,0 +1,184 @@
+                (external_key, username): Error message if there was an error
+            }
+
+        This function will look up program enrollments and users, and update the program enrollments
+        with the matching user. If the program enrollment has course enrollments, we will enroll the user into their
+        waiting program courses.
+
+        For each external_user_key:lms_username, if:
+            - The user is not found
+            - No enrollment is found for the given program and external_user_key
+            - The enrollment already has a user
+        An error message will be logged, and added to a dictionary of error messages keyed by (external_key, username).
+        The input will be skipped. All other inputs will be processed and enrollments updated, and then the function
+        will return the dictionary of error messages.
+
+        If there is an error while enrolling a user in a waiting program course enrollment, the error will be
+        logged, and added to the returned error dictionary, and we will roll back all transactions for that user so
+        that their db state will be the same as it was before this function was called, to prevent program enrollments
+        to be in a state where they have an LMS user but still have waiting course enrollments. All other inputs will
+        be processed normally.
+    """
+    errors = {}
+    program_enrollments = get_program_enrollments(program_uuid, external_keys_to_usernames.keys())
+    users = get_lms_users(external_keys_to_usernames.values())
+    for item in external_keys_to_usernames.items():
+        external_student_key, username = item
+
+        user = users.get(username)
+        error_message = None
+        if not user:
+            error_message = NO_LMS_USER_TPL.format(username)
+
+        program_enrollment = program_enrollments.get(external_student_key)
+        if not program_enrollment:
+            error_message = NO_PROGRAM_ENROLLMENT_TPL.format(
+                program_uuid=program_uuid,
+                external_student_key=external_student_key
+            )
+        elif program_enrollment.user:
+            error_message = get_existing_user_message(program_enrollment, user)
+
+        if error_message:
+            logger.warning(error_message)
+            errors[item] = error_message
+            continue
+
+        try:
+            with transaction.atomic():
+                link_program_enrollment_to_lms_user(program_enrollment, user)
+        except (CourseEnrollmentException, IntegrityError) as e:
+            # pdb.set_trace()
+            logger.exception(u"Rolling back all operations for {}:{}".format(
+                external_student_key,
+                username,
+            ))
+            error_message = type(e).__name__
+            if str(e):
+                error_message += ': '
+                error_message += str(e)
+            errors[item] = error_message
+
+            continue  # transaction rolled back
```

Comment:
Nit: I think `continue` is superfluous here, since it's the last line of your for-loop's body.